lib/mayaUsd/ufe/XformOpUtils.cpp

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "XformOpUtils.h"
+
+#include <pxr/usd/usdGeom/xformable.h>
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+template <bool INCLUSIVE>
+GfMatrix4d computeLocalTransform(
+    const std::vector<UsdGeomXformOp>&          ops,
+    std::vector<UsdGeomXformOp>::const_iterator endOp,
+    const UsdTimeCode&                          time)
+{
+    // If we want the op to be included, increment the end op iterator.
+    if (INCLUSIVE) {
+        TF_AXIOM(endOp != ops.end());
+        ++endOp;
+    }
+
+    // GetLocalTransformation() interface does not allow passing a begin and
+    // end iterator, so copy into an argument vector.
+    std::vector<UsdGeomXformOp> argOps(std::distance(ops.begin(), endOp));
+    argOps.assign(ops.begin(), endOp);
+
+    GfMatrix4d m(1);
+    if (!UsdGeomXformable::GetLocalTransformation(&m, argOps, time)) {
+        TF_FATAL_ERROR("Local transformation computation failed.");
+    }
+
+    return m;
+}
+
+template <bool INCLUSIVE>
+GfMatrix4d
+computePrimLocalTransform(const UsdPrim& prim, const UsdGeomXformOp& op, const UsdTimeCode& time)
```

Comment:
I would still call this computeLocalTransform(), and rely on function overloading to resolve to the right call.  To me, computePrimLocalTransform() has the meaning that we'll compute the local transform for the whole prim, which is not what we're doing.

Reply:
The reason I gave it a different name is that the using declaration below cannot work for overloaded templates. And, even it the compiler did find them ambiguous, those using declaration would have to have different names since you cannot overload a using declaration.

OTOH, if you have a better suggestion for the name I can change it. I named it Prim because it acts on a prim. (I also don't understand the comment about the whole prim. The XForm are taken from the prim, are they not affecting the prim?)

Reply:
What could be done is make the names that are declared in the "using" statements be real functions, which could then be overloaded, and only declare the templates, with different names, inside the cpp file, hidden from view.

Reply:
computeLocalTransform() in this case is computing the local transform up to (and optionally including) a given transform op in the prim's transform stack, not the local transform of the complete transform stack.

The template is an implementation detail to avoid code duplication, so as long as you find a way to keep that idea, I'm fine with whatever works.